File: SudokuKMP/iosApp/iosApp/ViewModels/GameViewModel.swift
import Foundation
import Combine

struct GameUIState {
    var grid: SudokuGrid = emptyGrid()
    var selectedCell: (row: Int, col: Int)? = nil
    var difficulty: Difficulty = .easy
    var isCompleted: Bool = false
    var elapsedSeconds: Int = 0
    var mistakes: Int = 0
    var maxMistakes: Int = 3
    var isSolverActive: Bool = false
    var solverFillingCell: (Int, Int)? = nil
    var solverSteps: [SolverStep] = []
    var currentSolverReasoning: String = ""
    var isGameOver: Bool = false
    var showCompletionDialog: Bool = false
}

@MainActor
class GameViewModel: ObservableObject {
    @Published var state = GameUIState()

    private let validator = SudokuValidator()
    private lazy var generator = PuzzleGenerator(validator: validator)
    private lazy var solver = SudokuSolver(validator: validator)

    private var timerTask: Task<Void, Never>?
    private var solverTask: Task<Void, Never>?

    private let userDefaults = UserDefaults.standard

    init() {
        startNewGame()
    }

    func startNewGame(difficulty: Difficulty? = nil) {
        stopSolver()
        timerTask?.cancel()
        let diff = difficulty ?? state.difficulty
        let puzzle = generator.generatePuzzle(difficulty: diff)
        state = GameUIState(
            grid: puzzle,
            difficulty: diff,
            elapsedSeconds: 0
        )
        startTimer()
        recordGamePlayed()
        updateStreak()
    }

    func selectCell(row: Int, col: Int) {
        guard !state.isCompleted && !state.isGameOver else { return }
        state.selectedCell = (row, col)
    }

    func enterNumber(_ num: Int) {
        guard !state.isCompleted, !state.isGameOver,
              let sel = state.selectedCell else { return }
        let cell = state.grid[sel.row][sel.col]
        guard !cell.isInitial else { return }

        if num != 0 {
            var intGrid = toIntGrid(state.grid)
            intGrid[sel.row][sel.col] = 0
            if !validator.isValidMove(grid: intGrid, row: sel.row, col: sel.col, num: num) {
                state.mistakes += 1
                if state.mistakes >= state.maxMistakes {
                    timerTask?.cancel()
                    state.isGameOver = true
                }
                return
            }
        }
        state.grid[sel.row][sel.col].value = num
        checkCompletion()
    }

    func eraseCell() {
        guard let sel = state.selectedCell,
              !state.grid[sel.row][sel.col].isInitial else { return }
        state.grid[sel.row][sel.col].value = 0
    }

    func startSolver(stepDelayMs: Int = 1500, maxCells: Int = 3) {
        if state.isSolverActive { stopSolver(); return }
        state.isSolverActive = true
        state.solverSteps = []

        solverTask = Task {
            var filled = 0
            while filled < maxCells {
                guard let step = solver.findNextStep(grid: state.grid) else { break }
                state.solverFillingCell = (step.row, step.col)
                state.currentSolverReasoning = step.reasoning

                try? await Task.sleep(nanoseconds: UInt64(stepDelayMs) * 1_000_000)
                guard !Task.isCancelled else { break }

                state.grid[step.row][step.col].value = step.value
                state.grid[step.row][step.col].isSolverFilled = true
                state.solverFillingCell = nil
                state.solverSteps.append(step)
                checkCompletion()

                filled += 1
                if state.isCompleted { break }
            }
            state.isSolverActive = false
        }
    }

    func stopSolver() {
        solverTask?.cancel()
        state.isSolverActive = false
        state.solverFillingCell = nil
    }

    func dismissCompletionDialog() { state.showCompletionDialog = false }

    // MARK: - Private

    private func startTimer() {
        timerTask?.cancel()
        timerTask = Task {
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 1_000_000_000)
                guard !Task.isCancelled else { break }
                state.elapsedSeconds += 1
            }
        }
    }

    private func checkCompletion() {
        if validator.isComplete(grid: state.grid) && validator.isValidSolution(grid: state.grid) {
            timerTask?.cancel()
            stopSolver()
            state.isCompleted = true
            state.showCompletionDialog = true
            recordWin()
        }
    }

    // MARK: - Persistence

    private func recordGamePlayed() {
        let count = userDefaults.integer(forKey: "games_played")
        userDefaults.set(count + 1, forKey: "games_played")
    }

    private func recordWin() {
        let won = userDefaults.integer(forKey: "games_won")
        userDefaults.set(won + 1, forKey: "games_won")
        let total = userDefaults.integer(forKey: "total_time")
        userDefaults.set(total + state.elapsedSeconds, forKey: "total_time")
        let bestKey: String
        switch state.difficulty {
        case .easy: bestKey = "best_easy"
        case .medium: bestKey = "best_medium"
        case .hard: bestKey = "best_hard"
        }
        let current = userDefaults.integer(forKey: bestKey)
        if current == 0 || state.elapsedSeconds < current {
            userDefaults.set(state.elapsedSeconds, forKey: bestKey)
        }
    }

    private func updateStreak() {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        let today = formatter.string(from: Date())
        let yesterday = formatter.string(from: Calendar.current.date(byAdding: .day, value: -1, to: Date())!)
        let lastPlayed = userDefaults.string(forKey: "last_played") ?? ""
        let streak = userDefaults.integer(forKey: "streak")
        let newStreak: Int
        if lastPlayed == today { newStreak = streak }
        else if lastPlayed == yesterday { newStreak = streak + 1 }
        else if lastPlayed.isEmpty { newStreak = 1 }
        else { newStreak = 1 }
        userDefaults.set(newStreak, forKey: "streak")
        userDefaults.set(today, forKey: "last_played")
    }
}

File: SudokuKMP/androidApp/src/main/kotlin/com/nextjedi/sudokustreak/android/viewmodel/GameViewModel.kt
package com.nextjedi.sudokustreak.android.viewmodel

import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.intPreferencesKey
import androidx.datastore.preferences.core.longPreferencesKey
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.nextjedi.sudokustreak.engine.PuzzleGenerator
import com.nextjedi.sudokustreak.engine.SudokuSolver
import com.nextjedi.sudokustreak.engine.SudokuValidator
import com.nextjedi.sudokustreak.model.Difficulty
import com.nextjedi.sudokustreak.model.SolverStep
import com.nextjedi.sudokustreak.model.SudokuCell
import com.nextjedi.sudokustreak.model.SudokuGrid
import com.nextjedi.sudokustreak.model.emptyGrid
import com.nextjedi.sudokustreak.model.toIntGrid
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

data class GameUiState(
    val grid: SudokuGrid = emptyGrid(),
    val selectedCell: Pair<Int, Int>? = null,
    val difficulty: Difficulty = Difficulty.EASY,
    val isCompleted: Boolean = false,
    val elapsedSeconds: Int = 0,
    val mistakes: Int = 0,
    val maxMistakes: Int = 3,
    val isSolverActive: Boolean = false,
    val solverHintCells: List<Pair<Int, Int>> = emptyList(),
    val solverFillingCell: Pair<Int, Int>? = null,
    val solverSteps: List<SolverStep> = emptyList(),
    val currentSolverReasoning: String = "",
    val isGameOver: Boolean = false,
    val showCompletionDialog: Boolean = false
)

class GameViewModel(private val dataStore: DataStore<Preferences>) : ViewModel() {

    private val validator = SudokuValidator()
    private val generator = PuzzleGenerator(validator)
    private val solver = SudokuSolver(validator)

    private val _uiState = MutableStateFlow(GameUiState())
    val uiState: StateFlow<GameUiState> = _uiState.asStateFlow()

    private var timerJob: Job? = null
    private var solverJob: Job? = null

    companion object {
        private val KEY_GAMES_PLAYED = intPreferencesKey("games_played")
        private val KEY_GAMES_WON = intPreferencesKey("games_won")
        private val KEY_TOTAL_TIME = longPreferencesKey("total_time")
        private val KEY_BEST_EASY = longPreferencesKey("best_easy")
        private val KEY_BEST_MEDIUM = longPreferencesKey("best_medium")
        private val KEY_BEST_HARD = longPreferencesKey("best_hard")
        private val KEY_STREAK = intPreferencesKey("streak")
        private val KEY_LAST_PLAYED = stringPreferencesKey("last_played")
    }

    fun startNewGame(difficulty: Difficulty = _uiState.value.difficulty) {
        stopSolver()
        timerJob?.cancel()

        val puzzle = generator.generatePuzzle(difficulty)
        _uiState.update {
            GameUiState(
                grid = puzzle,
                difficulty = difficulty,
                elapsedSeconds = 0
            )
        }

        startTimer()
        incrementGamesPlayed()
        updateStreak()
    }

    fun selectCell(row: Int, col: Int) {
        if (_uiState.value.isCompleted || _uiState.value.isGameOver) return
        _uiState.update { it.copy(selectedCell = Pair(row, col)) }
    }

    fun enterNumber(num: Int) {
        val state = _uiState.value
        if (state.isCompleted || state.isGameOver || state.selectedCell == null) return

        val (row, col) = state.selectedCell
        val cell = state.grid[row][col]
        if (cell.isInitial) return

        if (num != 0) {
            val intGrid = state.grid.toIntGrid()
            intGrid[row][col] = 0 // temporarily clear to validate
            if (!validator.isValidMove(intGrid, row, col, num)) {
                val newMistakes = state.mistakes + 1
                _uiState.update { it.copy(mistakes = newMistakes) }
                if (newMistakes >= state.maxMistakes) {
                    timerJob?.cancel()
                    _uiState.update { it.copy(isGameOver = true) }
                }
                return
            }
        }

        val newGrid = state.grid.mapIndexed { r, row_ ->
            row_.mapIndexed { c, cell_ ->
                if (r == row && c == col) cell_.copy(value = num) else cell_
            }
        }

        _uiState.update { it.copy(grid = newGrid) }
        checkCompletion()
    }

    fun eraseCell() {
        val state = _uiState.value
        if (state.isCompleted || state.isGameOver || state.selectedCell == null) return
        val (row, col) = state.selectedCell
        if (state.grid[row][col].isInitial) return
        enterNumber(0)
    }

    fun startSolver(stepDelayMs: Long = 1500L, maxCells: Int = 3) {
        if (_uiState.value.isSolverActive) {
            stopSolver()
            return
        }
        _uiState.update { it.copy(isSolverActive = true, solverSteps = emptyList()) }
        solverJob = viewModelScope.launch {
            var cellsFilled = 0
            while (cellsFilled < maxCells) {
                val step = solver.findNextStep(_uiState.value.grid) ?: break
                _uiState.update { state ->
                    state.copy(
                        solverFillingCell = Pair(step.row, step.col),
                        currentSolverReasoning = step.reasoning
                    )
                }
                delay(stepDelayMs)
                val state = _uiState.value
                val newGrid = state.grid.mapIndexed { r, row ->
                    row.mapIndexed { c, cell ->
                        if (r == step.row && c == step.col)
                            cell.copy(value = step.value, isSolverFilled = true)
                        else cell
                    }
                }
                _uiState.update { it.copy(
                    grid = newGrid,
                    solverFillingCell = null,
                    solverSteps = it.solverSteps + step
                ) }
                cellsFilled++
                checkCompletion()
                if (_uiState.value.isCompleted) break
            }
            _uiState.update { it.copy(isSolverActive = false) }
        }
    }

    fun stopSolver() {
        solverJob?.cancel()
        _uiState.update { it.copy(isSolverActive = false, solverFillingCell = null) }
    }

    fun dismissCompletionDialog() {
        _uiState.update { it.copy(showCompletionDialog = false) }
    }

    private fun startTimer() {
        timerJob?.cancel()
        timerJob = viewModelScope.launch {
            while (true) {
                delay(1000L)
                _uiState.update { it.copy(elapsedSeconds = it.elapsedSeconds + 1) }
            }
        }
    }

    private fun checkCompletion() {
        val state = _uiState.value
        if (validator.isComplete(state.grid) && validator.isValidSolution(state.grid)) {
            timerJob?.cancel()
            stopSolver()
            _uiState.update { it.copy(isCompleted = true, showCompletionDialog = true) }
            recordWin()
        }
    }

    private fun incrementGamesPlayed() {
        viewModelScope.launch {
            dataStore.edit { prefs ->
                prefs[KEY_GAMES_PLAYED] = (prefs[KEY_GAMES_PLAYED] ?: 0) + 1
            }
        }
    }

    private fun recordWin() {
        viewModelScope.launch {
            val elapsed = _uiState.value.elapsedSeconds.toLong()
            val diffKey = when (_uiState.value.difficulty) {
                Difficulty.EASY -> KEY_BEST_EASY
                Difficulty.MEDIUM -> KEY_BEST_MEDIUM
                Difficulty.HARD -> KEY_BEST_HARD
            }
            dataStore.edit { prefs ->
                prefs[KEY_GAMES_WON] = (prefs[KEY_GAMES_WON] ?: 0) + 1
                prefs[KEY_TOTAL_TIME] = (prefs[KEY_TOTAL_TIME] ?: 0L) + elapsed
                val current = prefs[diffKey] ?: Long.MAX_VALUE
                if (elapsed < current) prefs[diffKey] = elapsed
            }
        }
    }

    private fun updateStreak() {
        viewModelScope.launch {
            val prefs = dataStore.data.first()
            val lastPlayed = prefs[KEY_LAST_PLAYED] ?: ""
            val today = todayString()
            val currentStreak = prefs[KEY_STREAK] ?: 0
            val newStreak = when {
                lastPlayed == today -> currentStreak
                lastPlayed == yesterdayString() -> currentStreak + 1
                lastPlayed.isEmpty() -> 1
                else -> 1
            }
            dataStore.edit {
                it[KEY_STREAK] = newStreak
                it[KEY_LAST_PLAYED] = today
            }
        }
    }

    private fun todayString(): String =
        SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(Date())

    private fun yesterdayString(): String =
        SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(Date(System.currentTimeMillis() - 86_400_000L))

    override fun onCleared() {
        super.onCleared()
        timerJob?.cancel()
        solverJob?.cancel()
    }
}
